### PR TITLE
Fixing dataset evaluation

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
Fixed a bug that caused the error when calculating results on datasets other than COCO (like VOC). This was caused by the absence of the ```test_cfg``` description in the ```evaluate``` method, which is used in the CocoDataset only. 
Thanks to Leonid Beynenson for reporting.

### CODE REVIEW

**Comment by druzhkov-paul** [comment]
run tests

**Comment by druzhkov-paul** [comment]
@GalyaZalesskaya `evaluate` methods that you've touched have no common interface, i.e. use different set of parameters. Why should you pass `test_cfg` to all of them then? Could you please point me to the place, where `evaluate` is called for different datasets.

**Comment by GalyaZalesskaya** [comment]
@druzhkov-paul Hi, in ```test.py``` script there is the line that calls ```evaluate``` method which works differently for different datasets.  
https://github.com/openvinotoolkit/mmdetection/blob/9f89fde69a6efd7a4307956df5ab18d7f7388df6/tools/test.py#L157
As far as I need ```test_cfg``` parameter in the CocoDataset evaluation method, I pass it in ```dataset.evaluate``` in the test.py. So, if it is called for the other dataset, it provoked an error because its ```evaluate``` method do not expect ```test_cfg``` parameter.

**Comment by druzhkov-paul** [comment]
Ok. I do understand the reason of the problem and the way you propose to solve it. But I'd like to have a little discussion to get an understanding if there's a better way to solve it.

The drawback that see in the proposed solution is that it affects *all* datasets, i.e. requires changes in many places and put extra constraints on the datasets API.

So, `dataset.evaluate` is called at test time
https://github.com/openvinotoolkit/mmdetection/blob/ote/tools/test.py#L153-L157
and during training:
https://github.com/openvinotoolkit/mmdetection/blob/ote/mmdet/apis/train.py#L139-L142

The latter uses `evaluation` section from config file. Why don't we simply put extra parameters for text detection metrics in there instead of keeping them in `test_cfg`? It'll make the changes local. Dataset-related part of a config file for WIDER-face will contain an extra parameter(s) under the `evaluation` section and that's it.

@GalyaZalesskaya What are your thoughts here?

**Comment by GalyaZalesskaya** [comment]
@druzhkov-paul Your solution is great and quite elegant, thank you! 
But test script don't use parameters in ```evaluation``` section by default. So I will need to pass it with ```--options``` parameter that leads to modification of the [readme](https://github.com/openvinotoolkit/training_extensions/blob/develop/pytorch_toolkit/object_detection/horizontal-text-detection/readme.md) with command examples in the [training-extensions rep](https://github.com/openvinotoolkit/training_extensions). 
Are my conclusions right or I missed something?

**Comment by druzhkov-paul** [comment]
> But test script don't use parameters in evaluation section by default

@GalyaZalesskaya Why don't we just use them then? Something like that:

```python
kwargs = cfg.get('evaluation', {})
kwargs.update({} if args.options is None else args.options)
kwargs['metric'] = args.eval
if args.format_only:
    dataset.format_results(outputs, **kwargs)
if args.eval:
    dataset.evaluate(outputs, **kwargs)
```

Does it make sense?

**Comment by GalyaZalesskaya** [comment]
@druzhkov-paul It really does! That is so great!

File: tools/test.py [hunk 1]
```diff
@@ -150,11 +150,14 @@ def main():
         if args.out:
             print(f'\nwriting results to {args.out}')
             mmcv.dump(outputs, args.out)
-        kwargs = {} if args.options is None else args.options
+        kwargs = cfg.get('evaluation', {})
+        kwargs.pop('interval', None)
```
Comment (GalyaZalesskaya) [inline tools/test.py:154]: @druzhkov-paul I did changes that you proposed. I remove ```interval``` key from ```kwargs``` as far as it used to define the frequency of evaluating and datasets did not know how to process it. Does it look nice?

Comment (druzhkov-paul) [reply tools/test.py:154]: Ok, cool. Could you please remove `gpu_collect` key (if any) as well. It's used for evaluation hook in a distributed setup:

https://github.com/openvinotoolkit/mmdetection/ote/mmdet/core/evaluation/eval_hooks.py#L53

Comment (druzhkov-paul) [reply tools/test.py:154]: Add rebase the branch, please.

File: tools/test.py [hunk 2]
```diff
@@ -150,11 +150,15 @@ def main():
...
+        kwargs.pop('gpu_collect', None)
+        kwargs.update({} if args.options is None else args.options)
+        kwargs['metric'] = args.eval
         if args.format_only:
             dataset.format_results(outputs, **kwargs)
         if args.eval:
-            dataset.evaluate(outputs, args.eval, test_cfg=cfg.test_cfg, **kwargs)
+            dataset.evaluate(outputs, **kwargs)
```
Comment (Ilya-Krylov) [inline tools/test.py:161]: it becomes a bit tricky to change `eval` args via CLI

File: tools/test.py [hunk 3]
```diff
@@ -150,11 +150,15 @@ def main():
         if args.out:
             print(f'\nwriting results to {args.out}')
             mmcv.dump(outputs, args.out)
-        kwargs = {} if args.options is None else args.options
+        kwargs = cfg.get('evaluation', {})
+        kwargs.pop('interval', None)
+        kwargs.pop('gpu_collect', None)
+        kwargs.update({} if args.options is None else args.options)
+        kwargs['metric'] = args.eval
```
Comment (Ilya-Krylov) [inline tools/test.py:157]: looks like everything from `evaluation` is popped either overridden.

**Comment by druzhkov-paul** [comment]
run tests

**Review by druzhkov-paul — APPROVED** [review]

File: tools/test.py (hunk 3 above)
Comment (Ilya-Krylov) [reply tools/test.py:157]: I would introduce metrics with parameters. I don't like an idea to simply put a `score_thr` for `f1` score into `evaluation`. It is not clear that `score_thr` corresponds to `f1`. It is hard to change `score_thr` via CLI. I would propose something like that https://github.com/openvinotoolkit/mmdetection/commit/32ecee264f3a4da69e9a3dc616ff5d25cecdbf24

**Review by Ilya-Krylov — CHANGES_REQUESTED** [review]
I would introduce metrics with parameters. I don't like an idea to simply put a `score_thr` for `f1` score into `evaluation`. It is not clear that `score_thr` corresponds to `f1`. It is hard to change `score_thr` via CLI. I would propose something like that https://github.com/openvinotoolkit/mmdetection/commit/32ecee264f3a4da69e9a3dc616ff5d25cecdbf24

**Comment by druzhkov-paul** [comment]
@Ilya-Krylov Let's compare these two variants face to face in terms of CLI usage:

```
python tools/test.py config.py checkpoint.pth --out results.pkl --eval f1@thr=0.1
```
vs
```
python tools/test.py config.py checkpoint.pth --out results.pkl --eval f1 --options thr=0.1
```

IMHO pretty much the same.

**Comment by Ilya-Krylov** [comment]
@druzhkov-paul Probably yes, but in my case it is clearer that `thr` is for `f1`, we already have `score_thr` in `test_cfg`. It does not require to define additional argument in dataset.evaluate(...). The metric is defined in the same way in both CLI and config file.

**Comment by druzhkov-paul** [comment]
> it is clearer that `thr` is for f1

This is true.

> we already have `score_thr` in `test_cfg`

What for? Is it still needed?

> It does not require to define additional argument in dataset.evaluate(...)

What do you propose to do with already existing arguments? How about merge conflicts with upstream?

**Comment by Ilya-Krylov** [comment]
>     we already have `score_thr` in `test_cfg`
> What for? Is it still needed?

`score_thr` is used to filter out too low-confident detections I guess (since it is value is low). To compute mAP and f1 simultaneously one need to have two thresholds or set first one to 0 or pretty low value.

>     It does not require to define additional argument in dataset.evaluate(...)
> What do you propose to do with already existing arguments? How about merge conflicts with upstream?

I don't see any big conflicts with upstream, do you? On the contrary, I suggest to remove `score_thr` (`test_cfg`) from Coco.evaluate() that was introduced by us and that does not exist in upstream. I don't propose any interface changes, just extension of existing functionality that to my knowledge does not conflict with upstream.

**Comment by druzhkov-paul** [comment]
Let me probably answer my own questions:

> Is it still needed?

No, it's not needed any more and should be removed if we agree on the change that Galina introduced in this PR. This is a metric evaluation option, not a test process option.

> What do you propose to do with already existing arguments?

You propose to leave all of them as is, but remove ad hoc `score_thr`.

So, Ok. I guess you can polish you solution: make it work for arbitrary options including list/tuple values and for all of the datasets; clearly define and handle priority for those parameters that are defined both as a function argument and a part of metric string name, etc. Then we'll be able to review and discuss it.

**Comment by Ilya-Krylov** [comment]
>     Is it still needed?
> No, it's not needed any more and should be removed if we agree on the change that Galina introduced in this PR. This is a metric evaluation option, not a test process option.

I think it should not be removed in scope of this PR at least. If we remove we might not be able to reproduce metrics values of pre-trained models from upstream. They do use `score_thr` and values are different from model to model.

>     What do you propose to do with already existing arguments?
> You propose to leave all of them as is, but remove ad hoc `score_thr`.

Exactly

> So, Ok. I guess you can polish you solution: make it work for arbitrary options including list/tuple values and for all of the datasets; clearly define and handle priority for those parameters that are defined both as a function argument and a part of metric string name, etc. Then we'll be able to review and discuss it.

This is extra work definitely and is not required for now. Proposed changed are enough to work with parametrized version of `f1` in case of Coco dataset. List and tuples and other stuff might never be needed in case of metric definition.

**Comment by druzhkov-paul** [comment]
> I think it should not be removed in scope of this PR at least.

If so, I did understand it wrong. Let's not touch it then.

> List and tuples and other stuff might never be needed in case of metric definition.

Meanwhile, they are [already here](https://github.com/openvinotoolkit/mmdetection/blob/ote/mmdet/datasets/coco.py#L337-L338).

**Comment by Ilya-Krylov** [comment]
> Meanwhile, they are [already here](https://github.com/openvinotoolkit/mmdetection/blob/ote/mmdet/datasets/coco.py#L337-L338).

We can keep them as they are, not changing interface, considering them as standard metrics for Coco dataset.

But parametrized naming can be used for custom metrics like `f1`, also we can consider proposed approach as special case for `f1` but not general solution.
To summarize advantages:
1. No changes in interface relatively to upstream
2. Clear relationship between metric name and threshold
3. No confusion between existing `score_thr` and new parameter in `evaluation`
4. The same definition in CLI and configuration file.

I agree it is not suitable for general case, for COCO metrics for example. But it don't think that it is needed.

**Review by Ilya-Krylov — APPROVED** [review]
@GalyaZalesskaya do whatever you like

**Comment by GalyaZalesskaya** [comment]
run tests

**Comment by GalyaZalesskaya** [comment]
@druzhkov-paul @Ilya-Krylov I like the modification with parameterized f-score.
It has the following benefits:
+ No repeated parameter in ```test_cfg``` and ```evaluation``` section
+ It allows not to pop arguments like ```interval``` and ```gpu_collect``` in ```kwargs``` which is got from the evaluation section
+ It allows quick threshold changes right from the terminal to find the best value (but it possible to make it other ways)
+ I think that it is great that parser of metric arguments is located in ```evaluate``` function of a certain dataset. It makes the common interface unchanged.

Also, I see the following drawbacks:
- I did not see this hint before, including upstream, so it looks (in my point of view) a little bit scary. And at least it will be necessary to clearly define the possibility of parametrized metrics in READMEs
- In case of metrics with multiple significant parameters its spelling and understanding became more difficult (metric=f1@thr=0.25,other=True,another=True)
- I need to test argument parser code in order to find some (if any) non-obvious errors

Now I have the stable version without a parametrized metric. I can rewrite it by today.

**Comment by Ilya-Krylov** [comment]
@GalyaZalesskaya I think we can merge what you agreed on with @druzhkov-paul
We can further discuss it when my text-spotting branch is ready for review.
